src/fsharp/service/FSharpCheckerResults.fsi

```diff
@@ -82,6 +82,15 @@ type public FSharpProjectOptions =
     /// Compute the project directory.
     member internal ProjectDirectory: string
 
+and [<NoComparison>] public FSharpReferencedProject =
+    internal
+    | FSharp of projectFileName: string * options: FSharpProjectOptions
+    | IL of projectFileName: string * stamp: DateTime * lazyData: Lazy<byte []>
```

Comment:
Could you make it an interface? We can't use Roslyn compilation that produces a byte array but would be happy to use an implementation that looks at R# symbols.

Reply:
The use of the byte array isn't because of the Roslyn compilation, it's our metadata reader.

There are currently two functions for opening an ILModuleReader:
```fsharp
val internal OpenILModuleReader: string -> ILReaderOptions -> ILModuleReader

val internal OpenILModuleReaderFromBytes: fileName:string -> assemblyContents: byte[] -> options: ILReaderOptions -> ILModuleReader
```
I'm using the latter at the moment. Using the former is a little tricky as it tries to look at the last write time of a file which in this case, we can't because it's in-memory.

I think we need a new function to open an `ILModuleReader` that doesn't require a byte array and doesn't require the file to be on-disk.

Reply:
What do you think about something like this?
```fsharp
type FSharpReferencedProject =
    | FSharp of projectFileName: string * options: FSharpProjectOptions
    | IL of projectFileName: string * metadata: IILMetadata

and IILMetadata =
    abstract Timestamp: DateTime
    abstract CreateReader: unit -> ILModuleReader
```

Reply:
An IILMetadata would be better, though I am hesitant to create an ILModuleReader directly like that as a public API, even though ILModuleReader is public.

I need to investigate whether or not we could use System.Reflection.PortableExecutable.PEReader.

Reply:
What makes me want to reuse this logic is changing the module reader directly via shim currently makes it impossible to distinguish module reads from projects and from scripts. Using `FSharpReferencedProject` would make it a better experience, but we only can provide a custom module reader, not a PEReader or a metadata stream. Could you consider other options, please?

Reply:
To make sure I understand, you already have a module reader and you just want to pass it to a FSharpReferencedProject?

Reply:
Yes, I currently create it in `IAssemblyReader.GetILModuleReader(filename: string, readerOptions: ILReaderOptions)` implementation that is used in the assembly reader shim.

Reply:
Ok, then I think it makes sense to allow you to create a `FSharpReferencedProject` with an `ILModuleReader`. I will make another PR.

Reply:
@TIHan Have you looked into adding this case by any chance? I'll be happy to work on implementing it if we agree on some design.